src/ucp/tag/tag_send.c

```diff
@@ -220,87 +220,34 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_send_nb,
-    ucp_request_t *req;
-    ucs_status_ptr_t ret;
-
-    UCP_CONTEXT_CHECK_FEATURE_FLAGS(ep->worker->context, UCP_FEATURE_TAG,
-                                    return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM));
-    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
-
-    ucs_trace_req("send_nb buffer %p count %zu tag %"PRIx64" to %s cb %p",
-                  buffer, count, tag, ucp_ep_peer_name(ep), cb);
-
-    status = UCS_PROFILE_CALL(ucp_tag_send_inline, ep, buffer, count,
-                              datatype, tag);
-    if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
-        ret = UCS_STATUS_PTR(status); /* UCS_OK also goes here */
-        goto out;
-    }
-
-    status = ucp_ep_resolve_dest_ep_ptr(ep, ucp_ep_config(ep)->tag.lane);
-    if (status != UCS_OK) {
-        ret = UCS_STATUS_PTR(status);
-        goto out;
-    }
-
-    req = ucp_request_get(ep->worker);
-    if (req == NULL) {
-        ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
-        goto out;
-    }
-
-    ucp_tag_send_req_init(req, ep, buffer, datatype, count, tag, 0);
+    ucp_request_param_t params = {
+        .op_attr_mask = UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FIELD_CALLBACK,
+        .datatype     = datatype,
+        .cb.send      = (ucp_send_nbx_callback_t)cb
+    };
 
-    ret = ucp_tag_send_req(req, count, &ucp_ep_config(ep)->tag.eager,
-                           ucp_ep_config(ep)->tag.rndv.rma_thresh,
-                           ucp_ep_config(ep)->tag.rndv.am_thresh,
-                           cb, ucp_ep_config(ep)->tag.proto, 1);
-out:
-    UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(ep->worker);
-    return ret;
+    return ucp_tag_send_nbx(ep, buffer, count, tag, &params);
 }
 
 UCS_PROFILE_FUNC(ucs_status_t, ucp_tag_send_nbr,
                  (ep, buffer, count, datatype, tag, request),
                  ucp_ep_h ep, const void *buffer, size_t count,
                  uintptr_t datatype, ucp_tag_t tag, void *request)
 {
-    ucp_request_t *req = (ucp_request_t *)request - 1;
-    ucs_status_t status;
-    ucs_status_ptr_t ret;
-
-    UCP_CONTEXT_CHECK_FEATURE_FLAGS(ep->worker->context, UCP_FEATURE_TAG,
-                                    return UCS_ERR_INVALID_PARAM);
-    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
-
-    ucs_trace_req("send_nbr buffer %p count %zu tag %"PRIx64" to %s req %p",
-                  buffer, count, tag, ucp_ep_peer_name(ep), request);
-
-    status = UCS_PROFILE_CALL(ucp_tag_send_inline, ep, buffer, count,
-                              datatype, tag);
-    if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
-        UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(ep->worker);
-        return status;
-    }
-
-    status = ucp_ep_resolve_dest_ep_ptr(ep, ucp_ep_config(ep)->tag.lane);
-    if (status != UCS_OK) {
-        UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(ep->worker);
-        return status;
+    ucp_request_param_t param = {
+        .op_attr_mask = UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FIELD_REQUEST,
```

Comment:
what is about `UCP_OP_ATTR_FLAG_FAST_CMPL`? Maybe leave it as it was before (with zcopy disabled)? It is not used by the customer anyway?

Reply:
maybe better to leave this func untouched then? Otherwise it's behavior is not correct now, it does not ignore zcopy protocols anymore.
@yosefe, wdyt?

Reply:
reverted NO_ZCOPY flag